### DISCUSSION TITLE
Efficiently re-use vjp_funcs using jax loops

### DISCUSSION THREAD

**Discussion by BaptisteDebes** [jax-ml/jax#23180] [Q&A]
Hi all !

I'm writting an experiment that involves some variant of TBPTT. Here #22725 I learnt how it was possible to re-use gradient computation. My issue is that I need to do such a thing through **long** sequences in a loop which leads to exploding compilation times.

Here is a basic example where we cache some vjp_funcs to re-use them at later iteration.
Compilation time grows **dramatically** for non-trivial step functions (here I keep it simple).

Here is an example (where compilation times doesn't explode but it gives an idea).

```python

import jax
import jax.numpy as jnp
from jax import jit

N = 100
D = 500

key = jax.random.PRNGKey(0)

# Define a simple function whose VJP we will compute
def simple_function(x, key):
    key, subkey = jax.random.split(key)
    matrix = jax.random.normal(subkey, (D, D))
    return jnp.dot((x ** 2), matrix)

# Define the function to be used in the loop
def step_fn(carry, x, subkey):
    vjp_fn = carry
    vjp_result = vjp_fn(jnp.ones_like(x))[0]  # Use a dummy cotangent of 1
    out, new_vjp_fn = jax.vjp(simple_function, x, subkey)
    return new_vjp_fn, (out, vjp_result)

@jax.jit
def full_loop(array):
    key = jax.random.PRNGKey(0)
    initial_x = jnp.ones((D,))
    _, initial_vjp_fn = jax.vjp(simple_function, initial_x, key)
    
    carry = initial_vjp_fn
    results = []
    
    for i, x in enumerate(array):
        print(i)  # To track progress
        key, subkey = jax.random.split(key)
        carry, result = step_fn(carry, x, subkey)
        results.append(result)
    
    return carry

# Example usage
array = jax.random.normal(key, (N, D))
results = full_loop(array)

print("Results of the full loop:")
print(results)

```

I see two solutions 

1. Jit only the loop step function
2. Use lax.scan for lax.fori_loop

But the first one is not very satisfactory since it implies we can never run that type of computation in jitted context
The second solution (minimal example below) leads to an error due to the carry changing

```python

import jax
import jax.numpy as jnp
import jax.lax as lax

N = 100
D = 500

key = jax.random.PRNGKey(0)

# Define a simple function whose VJP we will compute
def simple_function(x, key):
    key, subkey = jax.random.split(key)
    matrix = jax.random.normal(subkey, (D, D))
    return jnp.dot((x ** 2), matrix)

# Define the function to be used in the loop
def step_fn(carry, x):
    vjp_fn, key = carry
    vjp_result = vjp_fn(jnp.ones_like(x))[0]  # Use a dummy cotangent of 1
    key, subkey = jax.random.split(key)
    out, new_vjp_fn = jax.vjp(simple_function, x, subkey)
    return (new_vjp_fn, key), (out, vjp_result)

@jax.jit
def scan_loop(array):
    key = jax.random.PRNGKey(0)
    initial_x = jnp.ones((D,))
    _, initial_vjp_fn = jax.vjp(simple_function, initial_x, key)
    
    carry = (initial_vjp_fn, key)
    _, results = lax.scan(step_fn, carry, array)
    
    return results

# Example usage
array = jax.random.normal(key, (N, D))
results = scan_loop(array)

print("Results of the scan loop:")
print(results)
```

> Scanned function carry input and carry output must have the same pytree structure, but they differ:
>   * the input carry component carry[0] is a <class 'jax._src.tree_util.Partial'> with pytree metadata _HashableCallableShim(functools.partial(<function _vjp_pullback_wrapper at 0x15266c05bf60>, 'simple_function', [dtype('float32')], [(500,)], (PyTreeDef(*), PyTreeDef((*, *))))) but the corresponding component of the carry output is a <class 'jax._src.tree_util.Partial'> with pytree metadata _HashableCallableShim(functools.partial(<function _vjp_pullback_wrapper at 0x15266c05bf60>, 'simple_function', [dtype('float32')], [(500,)], (PyTreeDef(*), PyTreeDef((*, *))))), so the pytree node metadata does not match
> 
> Revise the scanned function so that its output is a pair where the first element has the same pytree structure as the first argument.

I understand that we shouldn't be able to change the function completely from one iteration to another because computation must be static. However, in this case, it seems to me each new vjp_func involves the same computation. It looks pretty _pure_ to me. It is the _the same but just different numbers_ (for lack of a better terminology). 

My question: is there absolutely no way to do this ? I'm open to any idea as long as it keeps the efficiency of caching the vjp and doesn't lead to exploding compilation times.

Thank you very much in advance !

**Comment by BaptisteDebes** [comment]
up !

**Comment by BaptisteDebes** [comment]
up

**Comment by BaptisteDebes** [comment]
up

**Comment by jakevdp** [comment]
Sorry for the lack of response here. The loop carry can only contain arrays, not functions. You should avoid returning a function from your loop carry.

**Reply by BaptisteDebes** [reply]
Hi ! Thank you for your reply !!! 

The issue is that I simply don't know how to re-factor the code to avoid that pattern. I'm aware of the limitations on the carry. Do you have an idea how I could achieve a similar effect ?

The end goal is to implement truncated backpropagation through time. While I could simply recompute what is necessary for the backward pass, I really would like to avoid having to run the forward pass more than once (as it is particularly costly in my case).

I also cannot afford to unroll a whole python loop and compile it all.

Thank you in advance !

**Reply by jakevdp** [reply]
Could you use the vjp function in the iteration where it is created, and pass the results via scan to the next iteration where they're needed?
